README.md
# ora
--
    import "gopkg.in/rana/ora.v4"

Package ora implements an Oracle database driver.

### Golang Oracle Database Driver ###

#### TL;DR; just use it ####

    import (
    	"database/sql"

    	_ "gopkg.in/rana/ora.v4"
    )

    func main() {
    	db, err := sql.Open("ora", "user/passw@host:port/sid")
    	defer db.Close()

    	// Set timeout (Go 1.8)
    	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
    	// Set prefetch count (Go 1.8)
    	ctx = ora.WithStmtCfg(ctx, ora.Cfg().StmtCfg.SetPrefetchCount(50000))
    	rows, err := db.QueryContext(ctx, "SELECT * FROM user_objects")
    	defer rows.Close()
    }

Call stored procedure with OUT parameters:

    import (
    	"gopkg.in/rana/ora.v4"
    )

    func main() {
    	env, srv, ses, err := ora.NewEnvSrvSes("user/passw@host:port/sid")
    	if err != nil {
    		log.Fatal(err)
    	}
    	defer env.Close()
    	defer srv.Close()
    	defer ses.Close()

    	var user string
    	if _, err = ses.PrepAndExe("BEGIN :1 := SYS_CONTEXT('USERENV', :2); END;", &res, "SESSION_USER"); err != nil {
    		log.Fatal(err)
    	}
    	log.Printf("user: %q", user)
    }


### Background

An Oracle database may be accessed through the
[database/sql](http://golang.org/pkg/database/sql) package or through the ora
package directly. database/sql offers connection pooling, thread safety, a
consistent API to multiple database technologies and a common set of Go types.
The ora package offers additional features including pointers, slices, nullable
types, numerics of various sizes, Oracle-specific types, Go return type
configuration, and Oracle abstractions such as environment, server and session.

The ora package is written with the Oracle Call Interface (OCI) C-language
libraries provided by Oracle. The OCI libraries are a standard for client
application communication and driver communication with Oracle databases.

The ora package has been verified to work with:

* Oracle Standard 11g (11.2.0.4.0), Linux x86_64 (RHEL6)

* Oracle Enterprise 12c (12.1.0.1.0), Windows 8.1 and AMD64.

### ---

* [Installation](https://github.com/rana/ora#installation)

* [Data Types](https://github.com/rana/ora#data-types)

* [SQL Placeholder Syntax](https://github.com/rana/ora#sql-placeholder-syntax)

* [Working With The Sql
Package](https://github.com/rana/ora#working-with-the-sql-package)

* [Working With The Oracle Package
Directly](https://github.com/rana/ora#working-with-the-oracle-package-directly)

* [Logging](https://github.com/rana/ora#logging)

* [Test Database Setup](https://github.com/rana/ora#test-database-setup)

* [Limitations](https://github.com/rana/ora#limitations)

* [License](https://github.com/rana/ora#license)

* [API Reference](http://godoc.org/github.com/rana/ora#pkg-index)

* [Examples](./examples)

### ---


### Installation

Minimum requirements are Go 1.3 with CGO enabled, a GCC C compiler, and Oracle
11g (11.2.0.4.0) or Oracle Instant Client (11.2.0.4.0).

Install Oracle or Oracle Instant Client.

Copy the [oci8.pc](contrib/oci8.pc) from the `contrib` folder (or the one for
your system, maybe tailored to your specific locations) to a folder in
`$PKG_CONFIG_PATH` or a system folder, such as

    cp -aL contrib/oci8.pc /usr/local/lib/pkgconfig/oci8.pc

The ora package has no external Go dependencies and is available on GitHub and
gopkg.in:

    go get gopkg.in/rana/ora.v4

*WARNING*: If you have Oracle Instant Client 11.2, you'll need to add "=lnnz11"
to the list of linked libs! Otherwise, you may encounter "undefined reference to
`nzosSCSP_SetCertSelectionParams' " errors. Oracle Instant Client 12.1 does not
need this.


### Data Types

The ora package supports all built-in Oracle data types. The supported Oracle
built-in data types are NUMBER, BINARY_DOUBLE, BINARY_FLOAT, FLOAT, DATE,
TIMESTAMP, TIMESTAMP WITH TIME ZONE, TIMESTAMP WITH LOCAL TIME ZONE, INTERVAL
YEAR TO MONTH, INTERVAL DAY TO SECOND, CHAR, NCHAR, VARCHAR, VARCHAR2,
NVARCHAR2, LONG, CLOB, NCLOB, BLOB, LONG RAW, RAW, ROWID and BFILE.
SYS_REFCURSOR is also supported.

Oracle does not provide a built-in boolean type. Oracle provides a single-byte
character type. A common practice is to define two single-byte characters which
represent true and false. The ora package adopts this approach. The oracle
package associates a Go bool value to a Go rune and sends and receives the rune
to a CHAR(1 BYTE) column or CHAR(1 CHAR) column.

The default false rune is zero '0'. The default true rune is one '1'. The bool
rune association may be configured or disabled when directly using the ora
package but not with the database/sql package.


### SQL Placeholder Syntax

Within a SQL string a placeholder may be specified to indicate where a Go
variable is placed. The SQL placeholder is an Oracle identifier, from 1 to 30
characters, prefixed with a colon (:). For example:

    // example Oracle placeholder uses a colon
    INSERT INTO T1 (C1) VALUES (:C1)

Placeholders within a SQL statement are bound by position. The actual name is
not used by the ora package driver e.g., placeholder names :c1, :1, or :xyz are
treated equally.


### LastInsertId

The `database/sql` package provides a LastInsertId method to return the last
inserted row's id. Oracle does not provide such functionality, but if you append
`... RETURNING col /*LastInsertId*/` to your SQL, then it will be presented as
LastInsertId. Note that you have to mark with a `/*LastInsertId*/` (case
insensitive) your `RETURNING` part, to allow ora to return the last column as
`LastInsertId()`. That column must fit in `int64`, though!


### Working With The Sql Package

You may access an Oracle database through the database/sql package. The
database/sql package offers a consistent API across different databases,
connection pooling, thread safety and a set of common Go types. database/sql
makes working with Oracle straight-forward.

The ora package implements interfaces in the database/sql/driver package
enabling database/sql to communicate with an Oracle database. Using database/sql
ensures you never have to call the ora package directly.

When using database/sql, the mapping between Go types and Oracle types may be
changed slightly. The database/sql package has strict expectations on Go return
types. The Go-to-Oracle type mapping for database/sql is:

    Go type		Oracle type

    int64		NUMBER°, BINARY_DOUBLE, BINARY_FLOAT, FLOAT

    float64		NUMBER¹, BINARY_DOUBLE, BINARY_FLOAT, FLOAT

    time.Time	TIMESTAMP, TIMESTAMP WITH TIME ZONE, TIMESTAMP WITH LOCAL TIME ZONE, DATE

    string		CHAR², NCHAR, VARCHAR, VARCHAR2, NVARCHAR2, LONG, CLOB, NCLOB

    bool		CHAR(1 BYTE)³, CHAR(1 CHAR)³

    []byte		BLOB, LONG RAW, RAW

    ° A select-list column defined as an Oracle NUMBER with zero scale e.g.,
    NUMBER(10,0) is returned as an int64. Either int64 or float64 may be inserted
    into a NUMBER column with zero scale. float64 insertion will have its fractional
    part truncated.

    ¹ A select-list column defined as an Oracle NUMBER with a scale greater than
    zero e.g., NUMBER(10,4) is returned as a float64. Either int64 or float64 may
    be inserted into a NUMBER column with a scale greater than zero.

    ² A select-list column defined as an Oracle CHAR with a length greater than 1
    e.g., CHAR(2 BYTE) or CHAR(2 CHAR) is returned as a string. A Go string of any
    length up to the column max length may be inserted into the CHAR column.

    ³ The Go bool value false is mapped to the zero rune '0'. The Go bool value
    true is mapped to the one rune '1'.

The "ora" driver is automatically registered for use with sql.Open, but you can
call ora.SetCfg to set the used configuration options including statement
configuration and Rset configuration.

        func init() {
    		drvCfg := ora.Cfg()
    		drvCfg.FalseRune = 'N'
    		drvCfg.TrueRune = 'Y'
    		drvCfg.TrueRune = 'Y'
    		ora.SetCfg(drvCfg)
    	}

When configuring the driver for use with database/sql, keep in mind that
database/sql has strict Go type-to-Oracle type mapping expectations.


### Working With The Oracle Package Directly

The ora package allows programming with pointers, slices, nullable types,
numerics of various sizes, Oracle-specific types, Go return type configuration,
and Oracle abstractions such as environment, server and session. When working
with the ora package directly, the API is slightly different than database/sql.

When using the ora package directly, the mapping between Go types and Oracle
types may be changed. The Go-to-Oracle type mapping for the ora package is:

    Go type				Oracle type

    int64, int32, int16, int8	NUMBER°, BINARY_DOUBLE, BINARY_FLOAT, FLOAT
    uint64, uint32, uint16, uint8
    Int64, Int32, Int16, Int8
    Uint64, Uint32, Uint16, Uint8
    *int64, *int32, *int16, *int8
    *uint64, *uint32, *uint16, *uint8
    []int64, []int32, []int16, []int8
    []uint64, []uint32, []uint16, []uint8
    []Int64, []Int32, []Int16, []Int8
    []Uint64, []Uint32, []Uint16, []Uint8

    float64, float32		NUMBER¹, BINARY_DOUBLE, BINARY_FLOAT, FLOAT
    Float64, Float32
    *float64, *float32
    []float64, []float32
    []Float64, []Float32

    time.Time			TIMESTAMP, TIMESTAMP WITH TIME ZONE,
    Time				TIMESTAMP WITH LOCAL TIME ZONE, DATE
    *time.Time
    []time.Time
    []Time

    string				CHAR², NCHAR, VARCHAR, VARCHAR2,
    String				NVARCHAR2, LONG, CLOB, NCLOB, ROWID
    *string
    []string
    []String

    bool				CHAR(1 BYTE)³, CHAR(1 CHAR)³
    Bool
    *bool
    []bool
    []Bool

    []byte, [][]byte	BLOB

    Lob, []Lob, *Lob	BLOB, CLOB

    Raw, []Raw			RAW, LONG RAW

    IntervalYM			INTERVAL MONTH TO YEAR
    []IntervalYM

    IntervalDS			INTERVAL DAY TO SECOND
    []IntervalDS

    Bfile				BFILE

    ° A select-list column defined as an Oracle NUMBER with zero scale e.g.,
    NUMBER(10,0) is returned as an int64 by default. Integer and floating point
    numerics may be inserted into a NUMBER column with zero scale. Inserting a
    floating point numeric will have its fractional part truncated.

    ¹ A select-list column defined as an Oracle NUMBER with a scale greater than
    zero e.g., NUMBER(10,4) is returned as a float64 by default. Integer and
    floating point numerics may be inserted into a NUMBER column with a scale
    greater than zero.

    ² A select-list column defined as an Oracle CHAR with a length greater than 1
    e.g., CHAR(2 BYTE) or CHAR(2 CHAR) is returned as a string. A Go string of any
    length up to the column max length may be inserted into the CHAR column.

    ³ The Go bool value false is mapped to the zero rune '0'. The Go bool value
    true is mapped to the one rune '1'.

An example of using the ora package directly:

    package main

    import (
    	"fmt"
    	"gopkg.in/rana/ora.v4"
    )

    func main() {
    	// example usage of the ora package driver
    	// connect to a server and open a session
    	env, err := ora.OpenEnv()
    	defer env.Close()
    	if err != nil {
    		panic(err)
    	}
    	srvCfg := ora.SrvCfg{Dblink: "orcl"}
    	srv, err := env.OpenSrv(&srvCfg)
    	defer srv.Close()
    	if err != nil {
    		panic(err)
    	}
    	sesCfg := ora.SesCfg{
    		Username: "test",
    		Password: "test",
    	}
    	ses, err := srv.OpenSes(sesCfg)
    	defer ses.Close()
    	if err != nil {
    		panic(err)
    	}

    	// create table
    	tableName := "t1"
    	stmtTbl, err := ses.Prep(fmt.Sprintf("CREATE TABLE %v "+
    		"(C1 NUMBER(19,0) GENERATED ALWAYS AS IDENTITY "+
    		"(START WITH 1 INCREMENT BY 1), C2 VARCHAR2(48 CHAR))", tableName))
    	defer stmtTbl.Close()
    	if err != nil {
    		panic(err)
    	}
    	rowsAffected, err := stmtTbl.Exe()
    	if err != nil {
    		panic(err)
    	}
    	fmt.Println(rowsAffected)

    	// begin first transaction
    	tx1, err := ses.StartTx()
    	if err != nil {
    		panic(err)
    	}

    	// insert record
    	var id uint64
    	str := "Go is expressive, concise, clean, and efficient."
    	stmtIns, err := ses.Prep(fmt.Sprintf(
    		"INSERT INTO %v (C2) VALUES (:C2) RETURNING C1 INTO :C1", tableName))
    	defer stmtIns.Close()
    	rowsAffected, err = stmtIns.Exe(str, &id)
    	if err != nil {
    		panic(err)
    	}
    	fmt.Println(rowsAffected)

    	// insert nullable String slice
    	a := make([]ora.String, 4)
    	a[0] = ora.String{Value: "Its concurrency mechanisms make it easy to"}
    	a[1] = ora.String{IsNull: true}
    	a[2] = ora.String{Value: "It's a fast, statically typed, compiled"}
    	a[3] = ora.String{Value: "One of Go's key design goals is code"}
    	stmtSliceIns, err := ses.Prep(fmt.Sprintf(
    		"INSERT INTO %v (C2) VALUES (:C2)", tableName))
    	defer stmtSliceIns.Close()
    	if err != nil {
    		panic(err)
    	}
    	rowsAffected, err = stmtSliceIns.Exe(a)
    	if err != nil {
    		panic(err)
    	}
    	fmt.Println(rowsAffected)

    	// fetch records
    	stmtQry, err := ses.Prep(fmt.Sprintf(
    		"SELECT C1, C2 FROM %v", tableName))
    	defer stmtQry.Close()
    	if err != nil {
    		panic(err)
    	}
    	rset, err := stmtQry.Qry()
    	if err != nil {
    		panic(err)
    	}
    	for rset.Next() {
    		fmt.Println(rset.Row[0], rset.Row[1])
    	}
    	if err := rset.Err(); err != nil {
    		panic(err)
    	}

    	// commit first transaction
    	err = tx1.Commit()
    	if err != nil {
    		panic(err)
    	}

    	// begin second transaction
    	tx2, err := ses.StartTx()
    	if err != nil {
    		panic(err)
    	}
    	// insert null String
    	nullableStr := ora.String{IsNull: true}
    	stmtTrans, err := ses.Prep(fmt.Sprintf(
    		"INSERT INTO %v (C2) VALUES (:C2)", tableName))
    	defer stmtTrans.Close()
    	if err != nil {
    		panic(err)
    	}
    	rowsAffected, err = stmtTrans.Exe(nullableStr)
    	if err != nil {
    		panic(err)
    	}
    	fmt.Println(rowsAffected)
    	// rollback second transaction
    	err = tx2.Rollback()
    	if err != nil {
    		panic(err)
    	}

    	// fetch and specify return type
    	stmtCount, err := ses.Prep(fmt.Sprintf(
    		"SELECT COUNT(C1) FROM %v WHERE C2 IS NULL", tableName), ora.U8)
    	defer stmtCount.Close()
    	if err != nil {
    		panic(err)
    	}
    	rset, err = stmtCount.Qry()
    	if err != nil {
    		panic(err)
    	}
    	row := rset.NextRow()
    	if row != nil {
    		fmt.Println(row[0])
    	}
    	if err := rset.Err(); err != nil {
    		panic(err)
    	}

    	// create stored procedure with sys_refcursor
    	stmtProcCreate, err := ses.Prep(fmt.Sprintf(
    		"CREATE OR REPLACE PROCEDURE PROC1(P1 OUT SYS_REFCURSOR) AS BEGIN "+
    			"OPEN P1 FOR SELECT C1, C2 FROM %v WHERE C1 > 2 ORDER BY C1; "+
    			"END PROC1;",
    		tableName))
    	defer stmtProcCreate.Close()
    	rowsAffected, err = stmtProcCreate.Exe()
    	if err != nil {
    		panic(err)
    	}

    	// call stored procedure
    	// pass *Rset to Exe to receive the results of a sys_refcursor
    	stmtProcCall, err := ses.Prep("CALL PROC1(:1)")
    	defer stmtProcCall.Close()
    	if err != nil {
    		panic(err)
    	}
    	procRset := &ora.Rset{}
    	rowsAffected, err = stmtProcCall.Exe(procRset)
    	if err != nil {
    		panic(err)
    	}
    	if procRset.IsOpen() {
    		for procRset.Next() {
    			fmt.Println(procRset.Row[0], procRset.Row[1])
    		}
    		if err := procRset.Err(); err != nil {
    			panic(err)
    		}
    		fmt.Println(procRset.Len())
    	}

    	// Output:
    	// 0
    	// 1
    	// 4
    	// 1 Go is expressive, concise, clean, and efficient.
    	// 2 Its concurrency mechanisms make it easy to
    	// 3
    	// 4 It's a fast, statically typed, compiled
    	// 5 One of Go's key design goals is code
    	// 1
    	// 1
    	// 3
    	// 4 It's a fast, statically typed, compiled
    	// 5 One of Go's key design goals is code
    	// 3
    }

Pointers may be used to capture out-bound values from a SQL statement such as an
insert or stored procedure call. For example, a numeric pointer captures an
identity value:

    // given:
    // CREATE TABLE T1 (
    // C1 NUMBER(19,0) GENERATED ALWAYS AS IDENTITY (START WITH 1 INCREMENT BY 1),
    // C2 VARCHAR2(48 CHAR))
    var id int64
    stmt, err = ses.Prep("INSERT INTO T1 (C2) VALUES ('GO') RETURNING C1 INTO :C1")
    stmt.Exe(&id)

A string pointer captures an out parameter from a stored procedure:

    // given:
    // CREATE OR REPLACE PROCEDURE PROC1 (P1 OUT VARCHAR2) AS BEGIN P1 := 'GO'; END PROC1;
    var str string
    stmt, err = ses.Prep("CALL PROC1(:1)")
    stmt.Exe(&str)

Slices may be used to insert multiple records with a single insert statement:

    // insert one million rows with single insert statement
    // given: CREATE TABLE T1 (C1 NUMBER)
    values := make([]int64, 1000000)
    for n, _ := range values {
    	values[n] = int64(n)
    }
    rowsAffected, err := ses.PrepAndExe("INSERT INTO T1 (C1) VALUES (:C1)", values)

The ora package provides nullable Go types to support DML operations such as
insert and select. The nullable Go types provided by the ora package are Int64,
Int32, Int16, Int8, Uint64, Uint32, Uint16, Uint8, Float64, Float32, Time,
IntervalYM, IntervalDS, String, Bool, Binary and Bfile. For example, you may
insert nullable Strings and select nullable Strings:

    // insert String slice
    // given: CREATE TABLE T1 (C1 VARCHAR2(48 CHAR))
    a := make([]ora.String, 5)
    a[0] = ora.String{Value: "Go is expressive, concise, clean, and efficient."}
    a[1] = ora.String{Value: "Its concurrency mechanisms make it easy to"}
    a[2] = ora.String{IsNull: true}
    a[3] = ora.String{Value: "It's a fast, statically typed, compiled"}
    a[4] = ora.String{Value: "One of Go's key design goals is code"}
    stmt, err = ses.Prep("INSERT INTO T1 (C1) VALUES (:C1)")
    stmt.Exe(a)

    // Specify OraS to Prep method to return ora.String values
    // fetch records
    stmt, err = ses.Prep("SELECT C1 FROM T1", OraS)
    rset, err := stmt.Qry()
    for rset.Next() {
    	fmt.Println(rset.Row[0])
    }

The `Stmt.Prep` method is variadic accepting zero or more `GoColumnType` which
define a Go return type for a select-list column. For example, a Prep call can
be configured to return an int64 and a nullable Int64 from the same column:

    // given: create table t1 (c1 number)
    stmt, err = ses.Prep("SELECT C1, C1 FROM T1", ora.I64, ora.OraI64)
    rset, err := stmt.Qry()
    for rset.Next() {
    	fmt.Println(rset.Row[0], rset.Row[1])
    }

Go numerics of various sizes are supported in DML operations. The ora package
supports int64, int32, int16, int8, uint64, uint32, uint16, uint8, float64 and
float32. For example, you may insert a uint16 and select numerics of various
sizes:

    // insert uint16
    // given: create table t1 (c1 number)
    value := uint16(9)
    stmt, err = ses.Prep("INSERT INTO T1 (C1) VALUES (:C1)")
    stmt.Exe(value)

    // select numerics of various sizes from the same column
    stmt, err = ses.Prep(
    	"SELECT C1, C1, C1, C1, C1, C1, C1, C1, C1, C1, FROM T1",
    	ora.I64, ora.I32, ora.I16, ora.I8, ora.U64, ora.U32, ora.U16, ora.U8,
    	ora.F64, ora.F32)
    rset, err := stmt.Qry()
    row := rset.NextRow()

If a non-nullable type is defined for a nullable column returning null, the Go
type's zero value is returned.

GoColumnTypes defined by the ora package are:

    Go type		GoColumnType

    int64		I64

    int32		I32

    int16		I16

    int8		I8

    uint64		U64

    uint32		U32

    uint16		U16

    uint8		U8

    float64		F64

    Int64		OraI64

    Int32		OraI32

    Int16		OraI16

    Int8		OraI8

    Uint64		OraU64

    Uint32		OraU32

    Uint16		OraU16

    Uint8		OraU8

    Float64		OraF64

    Float32		OraF32

    time.Time	T

    Time		OraT

    string		S

    String		OraS

    bool		B

    Bool		OraB

    []byte		Bin

    Raw			Bin

    Lob°		Bin or S

    default¹	D

    ° Lob will return binary data if the Oracle column is a BLOB; otherwise, Lob
      will return a string if the Oracle column is a CLOB.

    ¹ D represents a default mapping between a select-list column and a Go type.
    The default mapping is defined in RsetCfg.

When Stmt.Prep doesn't receive a GoColumnType, or receives an incorrect
GoColumnType, the default value defined in RsetCfg is used.

EnvCfg, SrvCfg, SesCfg, StmtCfg and RsetCfg are the main configuration structs.
EnvCfg configures aspects of an Env. SrvCfg configures aspects of a Srv. SesCfg
configures aspects of a Ses. StmtCfg configures aspects of a Stmt. RsetCfg
configures aspects of Rset. StmtCfg and RsetCfg have the most options to
configure. RsetCfg defines the default mapping between an Oracle select-list
column and a Go type. StmtCfg may be set in an EnvCfg, SrvCfg, SesCfg and
StmtCfg. RsetCfg may be set in a Stmt.

EnvCfg.StmtCfg, SrvCfg.StmtCfg, SesCfg.StmtCfg may optionally be specified to
configure a statement. If StmtCfg isn't specified default values are applied.
EnvCfg.StmtCfg, SrvCfg.StmtCfg, SesCfg.StmtCfg cascade to new descendent
structs. When ora.OpenEnv() is called a specified EnvCfg is used or a default
EnvCfg is created. Creating a Srv with env.OpenSrv() will use SrvCfg.StmtCfg if
it is specified; otherwise, EnvCfg.StmtCfg is copied by value to SrvCfg.StmtCfg.
Creating a Ses with srv.OpenSes() will use SesCfg.StmtCfg if it is specified;
otherwise, SrvCfg.StmtCfg is copied by value to SesCfg.StmtCfg. Creating a Stmt
with ses.Prep() will use SesCfg.StmtCfg if it is specified; otherwise, a new
StmtCfg with default values is set on the Stmt. Call Stmt.Cfg() to change a
Stmt's configuration.

An Env may contain multiple Srv. A Srv may contain multiple Ses. A Ses may
contain multiple Stmt. A Stmt may contain multiple Rset.

    // StmtCfg cascades to descendent structs
    // EnvCfg -> SrvCfg -> SesCfg -> StmtCfg -> RsetCfg

Setting a RsetCfg on a StmtCfg does not cascade through descendent structs.
Configuration of Stmt.Cfg takes effect prior to calls to Stmt.Exe and Stmt.Qry;
consequently, any updates to Stmt.Cfg after a call to Stmt.Exe or Stmt.Qry are
not observed.

One configuration scenario may be to set a server's select statements to return
nullable Go types by default:

    sc := &ora.SrvCfg{Dblink: "orcl"}
    sc.Dblink = "orcl"
    cfg := NewStmtCfg().
    	SetNumberInt(ora.OraI64).
    	SetNumberFloat(ora.OraF64).
    	SetBinaryDouble(ora.OraF64).
    	SetBinaryFloat(ora.OraF64).
    	SetFloat(ora.OraF64).
    	SetDate(ora.OraT).
    	SetTimestamp(ora.OraT).
    	SetTimestampTz(ora.OraT).
    	SetTimestampLtz(ora.OraT).
    	SetChar1(ora.OraS).
    	SetVarchar(ora.OraS).
    	SetLong(ora.OraS).
    	SetClob(ora.OraS).
    	SetBlob(ora.OraBin).
    	SetRaw(ora.OraBin).
    	SetLongRaw(ora.OraBin).
    	SetFetchLen(100).
    	SetLOBFetchLen(100)

CHANGELOG.md
# Changelog #

## master ##

## v4.1.16 ##

## v4.1.15 ##
  * Add (Un)Marshal{Text,JSON} to OCINum, for nicer marshaling to JSON.

## v4.1.14 ##
  * Allow specifying FetchLen and LOBFetchLen.
  * Add Rset.ColumnIndex helper function.

## v4.1.13 ##
  * Fix bindStirngSlice error of nullInds reuse without reset.

## v4.1.12 ##
  * Do not free ocistmt - panics (issue #222).

## v4.1.11 ##
  * Return ErrBadConn from ExecContext, QueryContext if the culprit is bad conn.
    Fixes #205.

## v4.1.10 ##
  * Fix missing OCI handle frees on error (issue #207).

## v4.1.9 ##
  * Fix context deadline issue 192.

## v4.1.7 ##
  * Fix missing Rset.Close in DrvQueryResult.Next.

## v4.1.6 ##
  * Fix memory consumption in defString.

## v4.1.5 ##
  * Allow Num, OraNum and OCINum as bind parameters.
  * Allow DRCP (Database Resident Connection Pooling).
  * Fix reconnection when disconnected from Oracle (return driver.ErrBadConn).
  * Fix LOB reading error happening only spuriously, under pressure - issue 159.

## v4.1.4 ##
  * Fix LOB reading (issue #159).
  * Remove ociErrorNL as it does locking, against its name and comment.
  * Remove all remaining ora.v3 traces.

## v4.1.3 ##
  * Open LOBs for reading only at the first read - this eliminates the error
    ORA-24804 when the Row contains more than one LOB.

## v4.1.2 ##
  * Fix LOB reading with ora.S by simplifying reading code and using AL32UTF8 explicitly.

## v4.0.1 ##
  * Add an ora.L type to specify ora.Lob as column type in Qry and Prep.

## v4.0.0 ##
  * Rewrite the tests to run parallel
  * Rewrite the tests to use subtests
  * Add RWMutex everywhere where needed to be -race-free
  * Change the configuration to use immutable structures (StmtCfg and RsetCfg).
    This means cfg.Set... methods returns a copy, does not change the original -
	you have to call drv/env/srv/ses/stmt/rset.SetCfg(cfg)!
  * Remove pooling from Con - that's already done in database/sql.
  * Allow setting column type to `S` for numeric columns
    (with `ora.SetCfg(ora.Cfg().SetNumberFloat(ora.S))`), too.
	This enables us you to `.Scan` into a `*string`.
  * Add Stmt.Parse for only parsing the query.
  * Fix open/close memory leak by adding more handle freeing - see #148.

## v3.8.0 ##
  * go1.8: support additional features, as requested by @kardianos in #127.
  * Change default column types for BLOB and CLOB: Bin and S, instead of D and D.
    With this change the driver will return a string/[]byte from [CB]LOB columns,
	instead of an io.ReadCloser.
  * Return driver.ErrBadConn for connection errors, to allow database/sql to reconnect.

## v3.7.4 ##
  * Add RTrimChar to StmtCfg, default true. This makes the strip of right padding of CHAR columns configurable.

## v3.7.3 ##
  * Fix a panic with Con.sysName.
  * Be more forgiving for empty/non-open/NULL result sets returned by Oracle.
  * Treat SQLT_INT columns returned by TimesTen just as SQLT_NUM.
  * Make *[]string, *[]{,u}int{16,32,64} and *[]{Ui,I}nt{16,32,64}, *[]Date bindable.
  * Use a bytesArena (sync.Pool-based []byte arena) for []byte allocations.

## v3.7.2 ##
  * Fix panic in Pool.Get - #118.
  * Make returned LOB colums obey the given GoColumnType, to be able to force the into string/[]byte - see #117.
  * Return nil for everywhere in def*.value, if possible (string is an exception, see #105, #106).

## v3.7.1 ##
  * Fix defDate NULL handling error resulting in reading garbage as Time.
  * Add Ses.SetAction to be able to set the session's Module and Action attributes.

## v3.7.0 ##
  * Add "C" field to Lob to allow setting bind parameters to CLOB (default is BLOB).
  * Add date.Date, an implementation for encode/decode Oracle 7 byte DATE format, to avoid the overhead of calling C.OCIDateTime... functions. And use it in Rsets.
  * Make ora.Date use date.Date.
  * Fix nested rsets (rset with rset as field).
  * Add a new Pool implementation, which has a simple 1-1 pairing between ses and srv - easier to use,
    and "automatically" correct for parallel execution.
  * get rid of NewSesCfg() and NewSrvCfg()

## v3.6 ##
  * Refactor def* (resultset columns) to be more common, and allow multiple row fetch.
  * Use multiple row fetch to speed SELECTs - closes issue #86.

## v3.5 ##
  * Modify default for CHAR(1) columns: use ora.S (string), NOT ora.B (bool).
  * Add connect " AS SYSDBA" functionality.

## v3.4 ##
  * Implement PL/SQL TABLE support to be able to call PL/SQL blocks/stored procedures with slices of simple types.
  * Add numberBigInt and numberBigFloat colum types.
    This allows tweaking the used column types for NUMBER columns with unknown scale/precision.
  * OCINum instead of OraNum - space saving and less C calls (implemented pure Go read/write of Oracle OCINumber, in num directory).

## v3.3 ##
  * Introduce new `Num` and `OraNum` data types and `N` Go Column Type to represent Oracle numbers fully, by exchanging them with Go as strings.
    Make `N` the default column type for numbers unrepresentable by int64 or float64 (more digits than 19 or 15).

## v3.2 ##
  * Rewrite for Go 1.6 cgo restrictions.

## v3.1 ##

  * Use the returned length info for several places when reading attributes - such as column names.
  * Use float64 for non-integer numeric column types.

## v3.0 ##

  * Big rewrite of handles to pool them properly.


gen.go
// +build ignore

// Copyright 2016 Tamás Gulácsi. All rights reserved.
// Use of this source code is governed by The MIT License
// found in the accompanying LICENSE file.

package main

import (
	"bytes"
	"io/ioutil"
	"log"
)

func main() {
	src := "defFloat64.go"
	b := readFile(src)
	dst := "defFloat32.go"
	log.Printf("%s => %s", src, dst)
	if err := ioutil.WriteFile(
		dst,
		bytes.Replace(b, []byte("64"), []byte("32"), -1),
		0644,
	); err != nil {
		log.Fatal(err)
	}

	src = "defInt64.go"
	b = readFile(src)
	for _, s := range []string{"8", "16", "32"} {
		dst = "defInt" + s + ".go"
		log.Printf("%s => %s", src, dst)
		if err := ioutil.WriteFile(
			dst,
			bytes.Replace(b, []byte("64"), []byte(s), -1),
			0644,
		); err != nil {
			log.Fatal(err)
		}
	}

	for _, pair := range [][2]string{
		{"OCI_NUMBER_SIGNED", "OCI_NUMBER_UNSIGNED"},
		{"int64", "uint64"},
		{"Int64", "Uint64"},
	} {
		b = bytes.Replace(b, []byte(pair[0]), []byte(pair[1]), -1)
	}
	dst = "defUint64.go"
	if err := ioutil.WriteFile(dst, b, 0644); err != nil {
		log.Fatal(err)
	}
	for _, s := range []string{"8", "16", "32"} {
		dst = "defUint" + s + ".go"
		log.Printf("%s => %s", src, dst)
		if err := ioutil.WriteFile(
			dst,
			bytes.Replace(b, []byte("64"), []byte(s), -1),
			0644,
		); err != nil {
			log.Fatal(err)
		}
	}

	for _, plus := range []string{"", "Ptr", "Slice"} {
		src = "bndFloat64" + plus + ".go"
		b := readFile(src)
		dst = "bndFloat32" + plus + ".go"
		log.Printf("%s => %s", src, dst)
		if err := ioutil.WriteFile(
			dst,
			bytes.Replace(
				bytes.Replace(b, []byte("64"), []byte("32"), -1),
				[]byte("floatSixtyFour("), []byte("float64("), -1),
			0644,
		); err != nil {
			log.Fatal(err)
		}

		src = "bndInt64" + plus + ".go"
		b = readFile(src)
		for _, s := range []string{"8", "16", "32"} {
			dst = "bndInt" + s + plus + ".go"
			log.Printf("%s => %s", src, dst)
			if err := ioutil.WriteFile(
				dst,
				bytes.Replace(
					bytes.Replace(b, []byte("64"), []byte(s), -1),
					[]byte("intSixtyFour("), []byte("int64("), -1),
				0644,
			); err != nil {
				log.Fatal(err)
			}
		}

		for _, pair := range [][2]string{
			{"OCI_NUMBER_SIGNED", "OCI_NUMBER_UNSIGNED"},
			{"int64", "uint64"},
			{"Int64", "Uint64"},
		} {
			b = bytes.Replace(b, []byte(pair[0]), []byte(pair[1]), -1)
		}
		dst = "bndUint64" + plus + ".go"
		if err := ioutil.WriteFile(dst, b, 0644); err != nil {
			log.Fatal(err)
		}
		for _, s := range []string{"8", "16", "32"} {
			dst = "bndUint" + s + plus + ".go"
			log.Printf("%s => %s", src, dst)
			if err := ioutil.WriteFile(
				dst,
				bytes.Replace(
					bytes.Replace(b, []byte("64"), []byte(s), -1),
					[]byte("intSixtyFour("), []byte("int64("), -1),
				0644,
			); err != nil {
				log.Fatal(err)
			}
		}
	}
}

func readFile(fn string) []byte {
	src, err := ioutil.ReadFile(fn)
	if err != nil {
		log.Fatal(err)
	}
	src = bytes.Replace(src, []byte("//go:generate "), []byte("// Generated from "+fn+" by "), -1)
	return src
}
